README.md
<h2 align="center">
  <a href="https://conversejs.org" target="_blank" rel="noopener">
    <img alt="Converse.js" src="https://github.com/conversejs/converse.js/blob/master/logo/readme.png" width="480">
  </a>
</h2>


[![XMPP Chat](https://inverse.chat/badge.svg?room=discuss@conference.conversejs.org)](https://inverse.chat/#converse/room?jid=discuss@conference.conversejs.org)
[![Travis](https://api.travis-ci.com/conversejs/converse.js.png?branch=master)](https://travis-ci.com/conversejs/converse.js)
[![Bountysource bounties](https://img.shields.io/bountysource/team/converse.js/activity.svg?maxAge=2592000)](https://www.bountysource.com/teams/converse.js/issues?tracker_ids=194169)
[![Translation status](https://hosted.weblate.org/widgets/conversejs/-/svg-badge.svg)](https://hosted.weblate.org/engage/conversejs/?utm_source=widget)

[Converse](https://conversejs.org) is a web based [XMPP/Jabber](https://xmpp.org) chat client.

You can either use it as a webchat app, or you can integrate it into your own website.

It's 100% client-side JavaScript, HTML and CSS and the only backend required
is a modern XMPP server.

Please support this project via [Patreon](https://www.patreon.com/jcbrand) or [Liberapay](https://liberapay.com/jcbrand)

## Demo

Converse is hosted and can be used at [https://conversejs.org](https://conversejs.org).

A demo showing anonymous login is available at [https://conversejs.org/demo/anonymous.html](https://conversejs.org/demo/anonymous.html)
and a demo which shows how you can embed a single chat room into a page is
avialable at [https://conversejs.org/demo/embedded.html](https://conversejs.org/demo/embedded.html).

## Documentation

The developer/integrator documentation can be found at [https://conversejs.org/docs/html](https://conversejs.org/docs/html).

You'll probably want to begin with the [quickstart guide](https://conversejs.org/docs/html/quickstart.html),
which shows you how to use the CDN (content delivery network) to quickly get a demo up and running.

## Converse modes

### Overlay

In overlay mode, Converse appears overlayed chats on top of the website.

![Screenshot of Converse in overlay mode](https://opkode.com/img/Screen-Shot-2018-09-19-at-19.15.16.png)

### Fullpage

In fullpage mode, Converse behaves like a single-page app that covers the whole browser viewport.

![Screenshot of Converse in fullpage mode](https://opkode.com/img/Screen-Shot-2018-09-19-at-19.16.46.png)

### Embedded

In embedded mode, Converse can be embedded into an element in the DOM.

![Screenshot of Converse in embedded mode](https://opkode.com/img/Converse-embedded.png)

## Features
-   Available as overlayed chat boxes or as a fullscreen application. See [inverse.chat](https://inverse.chat) for the fullscreen version.
-   Custom status messages
-   Desktop notifications
-   A [plugin architecture](https://conversejs.org/docs/html/plugin_development.html) based on [pluggable.js](https://conversejs.github.io/pluggable.js/)
-   Chat statuses (online, busy, away, offline)
-   Anonymous logins, see the [anonymous login demo](https://conversejs.org/demo/anonymous.html)
-   URL Previews (requires server support, for example [mod_ogp](https://modules.prosody.im/mod_ogp.html)
-   Translated into over 30 languages

### Supported XMPP Extensions

- [RFC-7395](https://tools.ietf.org/html/rfc7395) XMPP Subprotocol support for WebSocket
- [XEP-0004](https://xmpp.org/extensions/xep-0004.html) Data Forms
- [XEP-0030](https://xmpp.org/extensions/xep-0030.html) Service discovery
- [XEP-0045](https://xmpp.org/extensions/xep-0045.html) Multi-user chat rooms
- [XEP-0048](https://xmpp.org/extensions/xep-0048.html) Bookmarks
- [XEP-0050](https://xmpp.org/extensions/xep-0050.html) Ad-Hoc Commands
- [XEP-0054](https://xmpp.org/extensions/xep-0054.html) VCard-temp
- [XEP-0059](https://xmpp.org/extensions/xep-0059.html) Result Set Management
- [XEP-0060](https://xmpp.org/extensions/xep-0060.html) Publish-Subscribe (limited support)
- [XEP-0066](https://xmpp.org/extensions/xep-0066.html) Out of Band Data
- [XEP-0077](https://xmpp.org/extensions/xep-0077.html) In-band registration
- [XEP-0085](https://xmpp.org/extensions/xep-0085.html) Chat State Notifications
- [XEP-0115](https://xmpp.org/extensions/xep-0115.html) Entity Capabilities
- [XEP-0124](https://xmpp.org/extensions/xep-0124.html) Bidirectional-streams Over Synchronous HTTP (BOSH)
- [XEP-0144](https://xmpp.org/extensions/xep-0144.html) Roster item exchange
- [XEP-0156](https://xmpp.org/extensions/xep-0156.html) Discovering Alternative XMPP Connection Methods
- [XEP-0163](https://xmpp.org/extensions/xep-0163.html) Personal Eventing Protocol (limited support)
- [XEP-0184](https://xmpp.org/extensions/xep-0184.html) Message Receipt
- [XEP-0198](https://xmpp.org/extensions/xep-0198.html) Stream Management
- [XEP-0199](https://xmpp.org/extensions/xep-0199.html) XMPP Ping
- [XEP-0203](https://xmpp.org/extensions/xep-0203.html) Delayed Delivery
- [XEP-0206](https://xmpp.org/extensions/xep-0206.html) XMPP Over BOSH
- [XEP-0245](https://xmpp.org/extensions/xep-0245.html) The /me Command
- [XEP-0249](https://xmpp.org/extensions/xep-0249.html) Direct MUC Invitations
- [XEP-0280](https://xmpp.org/extensions/xep-0280.html) Message Carbons
- [XEP-0297](https://xmpp.org/extensions/xep-0297.html) Stanza Forwarding (limited support)
- [XEP-0308](https://xmpp.org/extensions/xep-0308.html) Last Message Correction
- [XEP-0313](https://xmpp.org/extensions/xep-0313.html) Message Archive Management
- [XEP-0316](https://xmpp.org/extensions/xep-0316.html) MUC Eventing protocol (limited support)
- [XEP-0317](https://xmpp.org/extensions/xep-0317.html) Hats (limited support)
- [XEP-0333](https://xmpp.org/extensions/xep-0333.html) Chat Markers (limited support)
- [XEP-0352](https://xmpp.org/extensions/xep-0352.html) Client State Indication
- [XEP-0357](https://xmpp.org/extensions/xep-0357.html) Push Notifications
- [XEP-0359](https://xmpp.org/extensions/xep-0359.html) Unique and Stable Stanza IDs
- [XEP-0363](https://xmpp.org/extensions/xep-0363.html) HTTP File Upload
- [XEP-0372](https://xmpp.org/extensions/xep-0372.html) References
- [XEP-0382](https://xmpp.org/extensions/xep-0382.html) Spoiler messages
- [XEP-0384](https://xmpp.org/extensions/xep-0384.html) OMEMO Encryption
- [XEP-0393](https://xmpp.org/extensions/xep-0393.html) Message styling
- [XEP-0422](https://xmpp.org/extensions/xep-0422.html) Message Fastening (limited support)
- [XEP-0424](https://xmpp.org/extensions/xep-0424.html) Message Retractions
- [XEP-0425](https://xmpp.org/extensions/xep-0425.html) Message Moderation
- [XEP-0437](https://xmpp.org/extensions/xep-0437.html) Room Activity Indicators
- [XEP-0453](https://xmpp.org/extensions/xep-0453.html) DOAP Usage in XMPP
- [XEP-0454](https://xmpp.org/extensions/xep-0454.html) OMEMO Media sharing

## Integration into other servers and frameworks

-   **[Alfresco](https://www.alfresco.com)**: [alfresco-js-chat-share](https://github.com/keensoft/alfresco-js-chat-share)
-   **[Django](https://www.djangoproject.com)**: [django-conversejs](https://pypi.python.org/pypi/django-conversejs) or [django-xmpp](https://github.com/fpytloun/django-xmpp)
-   **[Elgg](https://elgg.org)**: [plugin](https://elgg.org/plugins/2997196)
-   **[Friendica](https://friendi.ca)**: [converse](https://github.com/friendica/friendica-addons/tree/master/xmpp/converse)
-   **[Openfire](https://www.igniterealtime.org/projects/openfire/index.jsp)**: [inverse.jar](https://www.igniterealtime.org/projects/openfire/plugins.jsp)
-   **[Patternslib](http://patternslib.com)**: [patterns.converse](https://github.com/jcbrand/patterns.converse)
-   **[Plone](https://plone.com)**: [collective.converse](https://github.com/collective/collective.converse)
-   **[Prosody](https://prosody.im/)**: [mod_conversejs](https://modules.prosody.im/mod_conversejs.html)
-   **[Pàdé](https://www.igniterealtime.org/projects/pade/index.jsp)**: [Pàdé](https://www.igniterealtime.org/projects/pade/index.jsp)
-   **[Roundcube](https://roundcube.net)**: [roundcube-converse.js-xmpp-plugin](https://github.com/devurandom/roundcube-converse.js-xmpp-plugin)
-   **[Ruby on Rails](https://rubyonrails.org)**: [conversejs-rails](https://github.com/mikemarsian/conversejs-rails)
-   **[Tiki Wiki CMS Groupware](https://tiki.org)**: [built-in optional feature](https://doc.tiki.org/XMPP)
-   **[Wordpress](https://wordpress.org)**: [ConverseJS](https://wordpress.org/plugins/conversejs/)


## Tests

We use behavior-driven tests written with [jasmine.js](https://jasmine.github.io/).

Run `make check` to execute all the tests.

## Licence

`Converse.js` is released under the [Mozilla Public License (MPL)](https://www.mozilla.org/MPL/2.0/index.txt).

## Attribution

Emoji images are courtesy of [Twemoji](https://emojitwo.github.io/).

## Support

Issues can be logged on the [Github issue tracker](https://github.com/conversejs/converse.js/issues).

## Donations

A heartfelt thanks for everyone who has supported this project over the years.
Many people have contributed testing, bugfixes, features and corrections.

We accept donations via [Patreon](https://www.patreon.com/jcbrand) and [Liberapay](https://liberapay.com/jcbrand).

## Sponsors

<p>
  <a href="https://blokt.com?utm_source=conversejs" target="_blank" rel="noopener">
    <img alt="Blokt Crypto & Privacy" src="https://raw.githubusercontent.com/conversejs/converse.js/541613d1fea8aef364af00180f60e959162e5e4b/logo/blokt.png" width="200">
  </a>
</p>
<p>
  <a href="https://www.codefirst.co.uk?utm_source=conversejs" target="_blank" rel="noopener">
    <img alt="Codefirst" src="https://raw.githubusercontent.com/conversejs/converse.js/541613d1fea8aef364af00180f60e959162e5e4b/logo/codefirst.png" width="200">
  </a>
</p>
<p>
  <a href="https://www.keycdn.com?utm_source=conversejs" target="_blank" rel="noopener">
    <img alt="KeyCDN" src="https://raw.githubusercontent.com/conversejs/converse.js/541613d1fea8aef364af00180f60e959162e5e4b/logo/keycdn.png" width="200">
  </a>
</p>
<p>
  <a href="https://mailbox.org?utm_source=conversejs" target="_blank" rel="noopener">
    <img alt="Mailbox - Email and cloud storage" src="https://raw.githubusercontent.com/conversejs/converse.js/master/logo/mailbox.svg" width="120">
  </a>
</p>


CHANGES.md
# Changelog

## 8.0.0 (Unreleased)

- Requires Node 12 or above

### Issues and features

- #1083: Add support for XEP-0393 Message Styling
- #1182: Add support for XEP-0454 OMEMO Media sharing
- #1823: New config options [mam_request_all_pages](https://conversejs.org/docs/html/configuration.html#mam-request-all-pages)
- #2275: Allow punctuation to immediately precede a mention
- #2348: `auto_join_room` not showing the room in `fullscreen` `view_mode`.
- #2400: Fixes infinite loop bug when appending .png to allowed image urls
- #2409: Integrate App Badging API for unread messages
- #2464: New configuration setting [allow_url_history_change](https://conversejs.org/docs/html/configuration.html#allow-url-history-change)
- #2497: Bugfix /nick command is not working
- Add a Description Of A Project (DOAP) file
- Add ability to deregister nickname when closing a MUC by setting `auto_register_muc_nickname` to `'unregister'`.
- Add support for XEP-0437 Room Activity Indicators see [muc-subscribe-to-rai](https://conversejs.org/docs/html/configuration.html#muc-subscribe-to-rai)
- Add support for rendering unfurls via [mod_ogp](https://modules.prosody.im/mod_ogp.html)
- Bugfix: Connection protocol not updated based on XEP-0156 connection methods
- Bugfix: Use real JID in XEP-0372 references only when the MUC is non-anonymous
- Bugfix: `null` inserted by emoji picker and can't switch between skintones
- File structure reordering: All plugins are now in `./plugins` folders.
- Show a gap placeholder when there are gaps in the chat history. The user can click these to fill the gaps.
- Use the MUC stanza id when sending XEP-0333 markers
- Add support for pausing Gif images

### New configuration setings

- [allow_url_history_change](https://conversejs.org/docs/html/configuration.html#allow-url-history-change)
- [mam_request_all_pages](https://conversejs.org/docs/html/configuration.html#mam-request-all-pages)
- [muc_clear_messages_on_leave](https://conversejs.org/docs/html/configuration.html#muc-clear-messages-on-leave)
- [muc_send_markers_for_own_messages](https://conversejs.org/docs/html/configuration.html#muc-send-markers-for-own-messages)
- [muc_send_probes](https://conversejs.org/docs/html/configuration.html#allow-url-history-change)
- [muc_show_ogp_unfurls](https://conversejs.org/docs/html/configuration.html#muc-show-ogp-unfurls)
- [prune_messages_above](https://conversejs.org/docs/html/configuration.html#prune-messages-above)
- [pruning_behavior](https://conversejs.org/docs/html/configuration.html#pruning-behavior)
- [send_chat_markers](https://conversejs.org/docs/html/configuration.html#send-chat-markers)
- [show_tab_notifications](https://conversejs.org/docs/html/configuration.html#show-tab-notifications)

### New hooks and events

- New hook: [getMessageActionButtons](https://conversejs.org/docs/html/api/-_converse.html#event:getMessageActionButtons)
- New hook: [shouldNotifyOfGroupMessage](https://conversejs.org/docs/html/api/-_converse.html#event:shouldNotifyOfGroupMessage)
- New hook: [presenceConstructed](https://conversejs.org/docs/html/api/-_converse.html#event:presenceConstructed)
- New event: [historyPruned](https://conversejs.org/docs/html/api/-_converse.html#event:historyPruned)

### Breaking Changes

The [afterMessageBodyTransformed](https://conversejs.org/docs/html/api/-_converse.html#event:afterMessageBodyTransformed) and [beforeMessageBodyTransformed](https://conversejs.org/docs/html/api/-_converse.html#event:beforeMessageBodyTransformed) events now has a different signatures.
When leaving a MUC, the message history is deleted. This means that decrypted
OMEMO messages are gone and cannot be recovered on that device. See [muc_clear_messages_on_leave](https://conversejs.org/docs/html/configuration.html#muc-clear-messages-on-leave).

Removed events:
* `bookmarkViewsInitialized`
* `chatBoxInsertedIntoDOM`
* `contactStatusMessageChanged`
* `messageSend` (use `sendMessage` instead)
* `rosterGroupsFetched`

The `chatBoxClosed`, `chatBoxMaximized` and `chatBoxMinimized` events now have the `model` as
payload and not the `view`.

The `api.chatviews` and `api.roomviews` API groupings and methods have been
removed. The original "views" don't really exist anymore, the UI is now composed web
components which should update reactively based on changes made on the models.

## 7.0.6 (unreleased)

- #2500: Wrong assignment in OMEMO code

## 7.0.5 (2021-03-16)

- #2377: The @converse/headless NPM package is missing the dist directory, causing import errors
- #2396: @converse/headless wrongly depends on `CustomElement` from the view layer
- #2381: Converse does not reflect the browser language

## 7.0.4 (2020-12-09)

Bugfix: Set protocol after discovering connection methods

Otherwise if an endpoint was passed in with `converse.initialize`, then
that endpoint's protocol is (potentially wrongly) used.

## 7.0.3 (2020-12-02)

- Bugfix: `null` inserted by emoji picker and can't switch between skintones


## 7.0.2 (2020-11-23)

- Updated translations: de, nb, gl, tr
- Still more necessary files not included in the 7.0.1 NPM package, causing `Module not found` errors.
- #2337: Newly opened headline chat isn't hidden by default

## 7.0.1 (2020-11-19)

- Updated translations: de, es, fi, fr, gl, hu, nb_NO, oc, pl, pt, pt_BR, ru, tr, uk, zh
- #2328: Honor `use_system_emojis` in emoji-picker
- Not all template/*.js files were included in the 7.0.0 NPM package, causing:
  `Module not found: Error: Can't resolve 'templates/*.js'`

## 7.0.0 (2020-11-18)

*Note for plugin authors:*
configuration settings should now be accessed via `_converse.api.settings.get` and not directly on the `_converse` object.
Soon we'll deprecate the latter, so prepare now.

- #515 Add support for XEP-0050 Ad-Hoc commands
- #2231: add sort_by_query and remove sort_by_length
- #1313: Stylistic improvements to the send button
- #1481: MUC OMEMO: Error No record for device
- #1490: Busy-loop when fetching registration form fails

src/headless/plugins/markers/utils.js
import ChatMarker from './marker.js';
import log from '@converse/headless/log';
import { MARKER_TYPES } from './constants.js';
import { _converse, api, converse } from '@converse/headless/core.js';
import { getOpenPromise } from '@converse/openpromise';
import { initStorage } from '@converse/headless/utils/storage.js';

const { $msg, Strophe, u } = converse.env;

export function getMessageIdToMark (message) {
    if (message.get('type') === 'groupchat') {
        const muc_jid = Strophe.getBareJidFromJid(message.get('from'));
        return message.get(`stanza_id ${muc_jid}`) || message.get('msgid');
    }
    return message.get('msgid');
}

/**
 * Given a message being marked by a particular JID, add (or update) a
 * ChatMarker message.
 *
 * See [XEP-0333](https://xmpp.org/extensions/xep-0333.html)
 *
 * @param { _converse.ChatBox | _converse.ChatRoom } chat
 * @param { _converse.Message } message - The message being marked
 * @param { String } by_jid - The JID of the user who sent a marker
 * @param { ('received'|'displayed'|'acknowledged') } type - The type of chat marker being added
 * @returns { ChatMarker }
 */
export function addChatMarker (chat, message, by_jid, type='displayed') {
    if (chat.get('type') === _converse.CHATROOMS_TYPE) {
        if (
            by_jid !== _converse.bare_jid &&
            chat.occupants.length > api.settings.get('muc_chat_markers_limit')
        ) {
            // XXX: this might cause orphaned ChatMarker instances in the chat.
            // For example if muc_chat_markers_limit is 10, then as the MUC
            // grows to 11 users, we will no longer create new ChatMarkers and
            // the older ones won't be removed.
            return;
        }
    }

    if (!Object.keys(MARKER_TYPES).includes(type)) {
        throw new TypeError('Invalid XEP-0333 chat marker type');
    }

    const marked_message_id = getMessageIdToMark(message);
    const marked = chat.markers.get(marked_message_id);
    const marked_entry = {};
    marked_entry[by_jid] = type;

    if (marked) {
        const marked_by = Object.assign(marked.get('marked_by') || {}, marked_entry);
        marked.save({ marked_by });
        return marked;
    } else {
        // Update (and potentially remove) existing markers to remove `by_jid`
        const predicate = m => m instanceof ChatMarker && m.get('marked_by')?.[by_jid];
        chat.markers.findWhere(predicate)?.removeMarkerJID(by_jid);

        const marker_data = {
            'id': marked_message_id,
            'marked_by': marked_entry,
            'time': (new Date()).setMilliseconds((new Date(message.get('time'))).getMilliseconds()+1)
        }
        return chat.markers.add(new ChatMarker(marker_data));
    }
}


/**
 * Send out a XEP-0333 chat marker
 * @param { String } to_jid
 * @param { String } id - The id of the message being marked
 * @param { String } type - The marker type
 * @param { String } msg_type
 */
export function sendChatMarker (to_jid, id, type, msg_type) {
    const stanza = $msg({
        'from': _converse.connection.jid,
        'id': u.getUniqueId(),
        'to': to_jid,
        'type': msg_type ? msg_type : 'chat'
    }).c(type, {'xmlns': Strophe.NS.MARKERS, 'id': id});
    api.send(stanza);
}


/**
 * Finds the last eligible message and then sends a XEP-0333 chat marker for it.
 * @param { (_converse.ChatBox|_converse.ChatRoom) } chat
 * @param { ('received'|'displayed'|'acknowledged') } [type='displayed']
 * @param { Boolean } [force=false] - Whether a marker should be sent for the
 *  message, even if it didn't include a `markable` element.
 */
export function sendMarkerForLastMessage (chat, type='displayed', force=false) {
    const msgs = Array.from(chat.messages.models);
    msgs.reverse();
    const msg = msgs.find(m => force || m.get('is_markable'));
    msg && sendMarkerForMessage(msg, type, force) && addChatMarker(chat, msg, _converse.bare_jid);
}


/**
 * Given the passed in message object, send a XEP-0333 chat marker if appropriate.
 * @param { _converse.Message } msg
 * @param { ('received'|'displayed'|'acknowledged') } [type='displayed']
 * @param { Boolean } [force=false] - Whether a marker should be sent for the
 *  message, even if it didn't include a `markable` element.
 * @returns { Boolean } Returns `true` or `false` depending on whether the
 *  marker was actually sent out.
 */
export function sendMarkerForMessage (msg, type='displayed', force=false) {
    if (!msg || !api.settings.get('send_chat_markers').includes(type)) {
        return false;
    }
    if (msg?.get('is_markable') || force) {
        const from_jid = Strophe.getBareJidFromJid(msg.get('from'));
        sendChatMarker(from_jid, msg.get('msgid'), type, msg.get('type'));
        const field_name = `marked_${type}`;
        const marked = msg.get(field_name) || [];
        msg.save({field_name: [...marked, _converse.bare_jid]});
        return true;
    }
    return false;
}

/**
 * Given the passed in MUC message, send a XEP-0333 chat marker if appropriate.
 * @param { _converse.MUCMessage } msg
 * @param { ('received'|'displayed'|'acknowledged') } [type='displayed']
 * @returns { Boolean } Returns `true` or `false` depending on whether the
 *  marker was actually sent out.
 */
export function sendMarkerForMUCMessage (chat, msg, type='displayed') {
    if (!Object.keys(MARKER_TYPES).includes(type)) {
        throw new TypeError('Invalid XEP-0333 chat marker type');
    }

    if (!msg || !api.settings.get('send_chat_markers').includes(type)) {
        return false;
    }
    if (msg?.get('is_markable')) {
        const mid = getMessageIdToMark(msg);
        if (chat.markers.get(mid)?.get('marked_by')?.[_converse.bare_jid] ?? -1 > MARKER_TYPES[type]) {
            // Already marked, either by the same marker value or by a higher ranked one.
            // https://xmpp.org/extensions/xep-0333.html#format
            return false;
        }
        const key = `stanza_id ${chat.get('jid')}`;
        const id = msg.get(key);
        if (!id) {
            log.error(`Can't send marker for message without stanza ID: ${key}`);
            return false;
        }
        const from_jid = Strophe.getBareJidFromJid(msg.get('from'));
        sendChatMarker(from_jid, id, type, msg.get('type'));
        return true;
    }
    return false;
}


/**
 * Given a new unread message in a chat, see whether we should send out a chat
 * marker for it.
 * @param { (_converse.ChatBox|_converse.ChatRoom) } chat
 * @param { _converse.Message } message
 */
export function handleUnreadMessage (chat, message) {
    if (!message?.get('body') || !u.isNewMessage(message) || chat.isHidden()) {
        return
    }
    sendMarkerForMessage(message) && addChatMarker(chat, message, _converse.bare_jid);
}


/**
 * Given an incoming message's attributes, check whether its a chat marker, and
 * if so, create a {@link _converse.ChatMarker}.
 * @param { MessageAttributes } attrs
 * @returns { Boolean } Returns `true` if the attributes are from a marker, and `false` otherwise.
 */
export function handleChatMarker (data, handled) {
    const { attrs, model } = data;
    const to_bare_jid = Strophe.getBareJidFromJid(attrs.to);
    if (to_bare_jid !== _converse.bare_jid) {
        return handled;
    }
    if (attrs.marker_id) {
        const message = model.messages.findWhere({'msgid': attrs.marker_id});
        message && addChatMarker(model, message, message.get('from'), message.get('marker'))
        return true;
    }
    return handled;
}

/**
 * Handler for the `messageUpdate` event. The point here is to send a marker
 * once we receive a reflected MUC message.
 *
 * The `muc_send_markers_for_own_messages` setting determines whether a marker
 * will be sent out.
 *
 * We send chat markers for our own sent messages to prevent Prosody's
 * mod_muc_rai (which implements XEP-0437 Room Activity Indicators) from
 * sending activity indicators for our own messages.
 *
 * @param { _converse.ChatRoom } chat
 * @param { _converse.Message } message
 */
export function onMessageUpdated (chat, message) {
    if (chat.isHidden() || chat.get('type') !== _converse.CHATROOMS_TYPE || !api.settings.get('muc_send_markers_for_own_messages')) {
        return;
    }
    sendMarkerForMUCMessage(chat, message) && addChatMarker(chat, message, _converse.bare_jid);
}

export function onUnreadsCleared (chat) {
    if (chat.get('type') === _converse.CHATROOMS_TYPE) {
        if (chat.get('num_unread_general') > 0 || chat.get('num_unread') > 0 || chat.get('has_activity')) {
            const message = chat.messages.last();
            sendMarkerForMUCMessage(chat, message) && addChatMarker(chat, message, _converse.bare_jid);
        }
    } else {
        if (chat.get('num_unread') > 0) {
            const message = chat.messages.last();
            sendMarkerForMessage(message) && addChatMarker(chat, message, _converse.bare_jid);
        }
    }
}

/**
 * Creates a ChatMarkers collection, set it on the chat and fetch any cached markers
 * @param { (_converse.ChatBox|_converse.ChatRoom) }
 */
export function initChatMarkers (model) {
    model.markers = new _converse.ChatMarkers();
    const id = `converse.markers-${model.get('jid')}`;
    initStorage(model.markers, id);
    model.markers.fetched = getOpenPromise();
    const resolve = model.markers.fetched.resolve;
    model.markers.fetch({
        'add': true,
        'success':  resolve,
        'error': resolve
    });
}
